File: php/order/duplicateOrder.php
<?php /** @noinspection DuplicatedCode */

require_once '../vendor/autoload.php';
require_once '../constants.php';

// These are the steps for cloning an existing order and charging the customer for it.
// 1. duplicateOrder
// 2. updateOrder (if you wish to change any part of it)
// 3. processPayment to charge the customer.
//
// As a reminder, if you wish to create a new order from scratch, use the CheckoutApi.
// The OrderApi is for managing existing orders.


$order_api = ultracart\v2\api\OrderApi::usingApiKey(Constants::API_KEY);

$expansion = "items";   // for this example, we're going to change the items after we duplicate the order, so
                        // the only expansion properties we need are the items.
                        // See: https://www.ultracart.com/api/  for a list of all expansions.

// Step 1. Duplicate the order
$order_id_to_duplicate = 'DEMO-0009104436';
$api_response = $order_api->duplicateOrder($order_id_to_duplicate, $expansion);
$new_order = $api_response->getOrder();

// Step 2. Update the items.  I will create a new items array and assign it to the order to remove the old ones completely.
$items = array();
$item = new ultracart\v2\models\OrderItem();
$item->setMerchantItemId('simple_teapot');
$item->setQuantity(1);
$item->setDescription("A lovely teapot");
$item->setDistributionCenterCode('DFLT'); // where is this item shipping out of?

$cost = new \ultracart\v2\models\Currency();
$cost->setCurrencyCode('USD');
$cost->setValue(9.99);
$item->setCost($cost);

$weight = new \ultracart\v2\models\Weight();
$weight->setUom("OZ");
$weight->setValue(6);
$item->setWeight($weight);

$items[] = $item;
$new_order->setItems($items);
$update_response = $order_api->updateOrder($new_order, $new_order->getOrderId(), $expansion);

$updated_order = $update_response->getOrder();

// Step 3. process the payment.
// the request object below takes two optional arguments.
// The first is an amount if you wish to bill for an amount different from the order.  We do not.
// The second is card_verification_number_token, which is a token you can create by using our hosted fields to
// upload a CVV value.  This will create a token you may use here.  However, most merchants using the duplicate
// order method will be setting up an auto order for a customer.  Those will not make use of the CVV, so we're
// not including it here.  That is why the request object below is does not have any values set.
// For more info on hosted fields, see: https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377775/UltraCart+Hosted+Credit+Card+Fields
$process_payment_request = new \ultracart\v2\models\OrderProcessPaymentRequest();
$payment_response = $order_api->processPayment($new_order->getOrderId(), $process_payment_request);
$transaction_details = $payment_response->getPaymentTransaction(); // do whatever you wish with this.

echo '<html lang="en"><body><pre>';
echo 'New Order (after updated items):<br>';
var_dump($updated_order);
echo '<br>Payment Response:<br>';
var_dump($payment_response);
echo '</pre></body></html>';

File: ruby/order/duplicate_order.rb
# OrderApi.duplicate_order() does not accomplish much on its own.  The use-case for this method is to
# duplicate a customer's order and then charge them for it.  duplicate_order() does not charge the customer again.
#
# These are the steps for cloning an existing order and charging the customer for it.
# 1. duplicate_order
# 2. update_order (if you wish to change any part of it)
# 3. process_payment to charge the customer.
#
# As a reminder, if you wish to create a new order from scratch, use the CheckoutApi or ChannelPartnerApi.
# The OrderApi is for managing existing orders.

require_relative '../constants'
require 'ultracart_api'

order_api = UltracartClient::OrderApi.new_using_api_key(Constants::API_KEY)

# Step 1. Duplicate the order
expansion = '_expand=items'   # for this example, we're going to change the items after we duplicate the order, so
# the only expansion properties we need are the items.
# See: https://www.ultracart.com/api/  for a list of all expansions.

order_id_to_duplicate = 'DEMO-0009104436'
api_response = order_api.duplicate_order(order_id_to_duplicate, { '_expand' => expansion })
new_order = api_response.get_order

# Step 2. Update the items. Create a new items array and assign it to the order to remove the old ones completely.
items = []
item = UltracartClient::OrderItem.new
item.set_merchant_item_id('simple_teapot')
item.set_quantity(1)
item.set_description("A lovely teapot")
item.set_distribution_center_code('DFLT') # where is this item shipping out of?

cost = UltracartClient::Currency.new
cost.set_currency_code('USD')
cost.set_value(9.99)
item.set_cost(cost)

weight = UltracartClient::Weight.new
weight.set_uom("OZ")
weight.set_value(6)
item.set_weight(weight)

items << item
new_order.set_items(items)
update_response = order_api.update_order(new_order.get_order_id, new_order, { '_expand' => expansion })

updated_order = update_response.get_order

# Step 3. process the payment.
# the request object below takes two optional arguments.
# The first is an amount if you wish to bill for an amount different from the order.
# We do not bill differently in this example.
# The second is card_verification_number_token, which is a token you can create by using our hosted fields to
# upload a CVV value.  This will create a token you may use here.  However, most merchants using the duplicate
# order method will be setting up an auto order for a customer.  Those will not make use of the CVV, so we're
# not including it here.  That is why the request object below does not have any values set.
# For more info on hosted fields:
# See: https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377775/UltraCart+Hosted+Credit+Card+Fields
# See: https://github.com/UltraCart/sdk_samples/blob/master/hosted_fields/hosted_fields.html

process_payment_request = UltracartClient::OrderProcessPaymentRequest.new
payment_response = order_api.process_payment(new_order.get_order_id, process_payment_request)
transaction_details = payment_response.get_payment_transaction # do whatever you wish with this.

puts 'New Order (after updated items):'
p updated_order
puts '\nPayment Response:'
p payment_response